woocommerce-services.php

```diff
@@ -1025,6 +1026,104 @@ public function reports_tabs( $reports ) {
+		public function track_completed_order( $order ) {
+			// We need to recreate a temporary cart from the order
+			// so that we can create the same shipping packages from the cart
+			// to use to generate the correct cache key to return the cached
+			// shipping rates response.
+			$logger = $this->get_logger();
+
+			$temp_cart   = new WC_Cart();
+			$order_items = $order->get_items();
+
+			foreach ( $order_items as $item ) {
+				$product = wc_get_product( $item->get_product() );
+
+				$product_id           = $product->get_id();
+				$quantity             = $item->get_quantity();
+				$variation_id         = 0;
+				$variation_attributes = array();
+
+				if ( 'variation' === $product->get_type() ) {
+					$variation_id         = $product_id;
+					$product_id           = $product->get_parent_id();
+					$variation_attributes = $product->get_variation_attributes();
+				}
+
+				$temp_cart->add_to_cart( $product_id, $quantity, $variation_id, $variation_attributes );
+			}
+
+			$packages = $temp_cart->get_shipping_packages();
+			$temp_cart->empty_cart();
+			$package = $packages[0];
+			foreach ( $package['contents'] as $item_key => $item_details ) {
+				ksort( $package['contents'][ $item_key ] );
+			}
+			$package['rates'] = array();
+
+			$shipping_zone    = WC_Shipping_Zones::get_zone_matching_package( $package );
+			$shipping_methods = $shipping_zone->get_shipping_methods( true );
+
+			// We will fill these with the required parameters
+			// should we need to repeat the shipping request for
+			// each service as a fallback.
+			$services         = array();
+			$custom_boxes     = array();
+			$predefined_boxes = array();
+			$requested_rates  = array();
+
+			foreach ( $shipping_methods as $method ) {
+				try {
+					$handler = new WC_Connect_Shipping_Method( $method->get_instance_id() );
+
+					$service_settings       = $handler->get_service_settings();
+					$service_schema         = $handler->get_service_schema();
+					$service_settings_store = $handler->get_service_settings_store();
+					$cached_response_body   = $handler->get_cached_shipping_rates_response( $package );
+
+					$services[]         = array(
+						array(
+							'id'               => $service_schema->id,
+							'instance'         => $method->get_instance_id(),
+							'service_settings' => $service_settings,
+						),
+					);
+					$custom_boxes[]     = $service_settings_store->get_packages();
+					$predefined_boxes[] = $service_settings_store->get_predefined_packages_for_service( $service_schema->id );
+					$requested_rates[]  = $cached_response_body;
+				} catch ( Exception $e ) {
+					$logger->error(
+						sprintf(
+							'Unable to send subscription event for method: %s',
+							$e->getMessage()
+						),
+						__FUNCTION__
+					);
+					continue;
+				}
+			}
+
+			$api_client = $this->get_api_client();
+			$response   = $api_client->track_subscription_event( $services, $package, $custom_boxes, $predefined_boxes, $requested_rates );
```

Comment:
It seems unnecessary to send all this data to the connect server. We only need to know which services were used and even that doesn't seem necessary since they all have the same event type.

Reply:
Limiting the amount data seems like a good goal. I would vote we include the services data though so we have some flexibility with what we track in the future without updating the client (there's been some discussion about tracking/limiting usage for certain carriers, but not others)

Reply:
Right now we only have one event type for all live rates, but I imagine this will change pretty quickly, if we wanted to have separate subscription products for different carriers. All this logic for the associations between carriers, usage events, and subscription products is contained in the Connect Server, so my plan from the start has basically been to send the rates there and let the server figure it out.

The reason why I'm sending the additional information about the shipping packages has almost entirely been my idea and I'm not sure how necessary it really is. The additional package information is only used as a fallback if the cached rates are not present for some reason. In this case, the server just repeats the shipping rates request to get the rates and then continues as normal. The only ways I can think that this would happen is if a user crippled the entire checkout by disabling cookies (though, I think the transient might even survive this, since it is database backed) or if a user took more than two hours to complete their checkout and the transient expired (which I think might generally break the entire checkout as well).

This is basically a fallback for a (hopefully) thin edge case, so if there are enough voices telling me that it's not worth worrying about, I will be content to sleep easy and disregard it. However, it doesn't seem like there's too much extra cost having it there though, other than the size of the HTTP request package to the server, since I already need to generate the shipping packages to try to get the cached rates out of the transient, as described above somewhere.

Reply:
I can't imagine how it could ever happen. The cart must generate the shipping prices during checkout and in the same request we are triggering this tracking method so there is no time at all for it to expire. We probably don't even need to use a transient and generate the caching key. We could have the shipping method call a add_tracking_event_data() which would set a property on this class. Then during the checkout hook we just need to read the tracking events that need to be sent.

Reply:
> We could have the shipping method call a add_tracking_event_data() which would set a property on this class.

I like this idea. Let me try this out tomorrow morning and I'll strip out the rates request fallback while I'm at it as well and report back.

Reply:
@superdav42 I've just removed this extra information, so now I am only sending the cached rates and the shipping method services information with the request, removing all the extraneous shipping package information.